src/compiler/emitter.ts

```diff
@@ -118,34 +124,73 @@ namespace ts {
             onSetSourceFile: setSourceFile,
         });
 
+        // Setup and perform the transformation to retrieve declarations from the input files
+        let declarationTransform: TransformationResult<SourceFile | Bundle>;
+        let declarationPrinter: Printer;
+        if (emitOnlyDtsFiles || compilerOptions.declaration) {
+            const nonJsFiles = filter(sourceFiles, isSourceFileNotJavaScript);
+            const inputListOrBundle = (compilerOptions.outFile || compilerOptions.out) ? [createBundle(nonJsFiles)] : nonJsFiles;
+            declarationTransform = transformNodes(resolver, host, compilerOptions, inputListOrBundle, [transformDeclarations], /*allowDtsFiles*/ false);
+            declarationPrinter = createPrinter({ ...compilerOptions, onlyPrintJsDocStyle: true } as PrinterOptions, {
+                // resolver hooks
+                hasGlobalName: resolver.hasGlobalName,
+
+                // transform hooks
+                onEmitNode: declarationTransform.emitNodeWithNotification,
+                substituteNode: declarationTransform.substituteNode,
+            });
+        }
+
         // Emit each output file
         performance.mark("beforePrint");
         forEachEmittedFile(host, emitSourceFileOrBundle, transform.transformed, emitOnlyDtsFiles);
         performance.measure("printTime", "beforePrint");
 
         // Clean up emit nodes on parse tree
         transform.dispose();
+        if (declarationTransform) declarationTransform.dispose();

```

Comment:
hmmm I implemented it that way initially, but it changes the order we emit files in (instead of all outputs for the same file being one after another, dts outputs come after everything else), which affects almost every baseline, as it happens. It would be better to make it "safe" to run multiple transforms at the same time, since nothing can enforce the invariant that you don't do so.

Reply:
Making it "safe" might necessitate additional indirection for a lot of utility functions that interact with `emitNode`. Perhaps it would be better to transform and print each source file one at a time for multi-file emit. That could also cut back on the amount of memory we need to hold onto between each file as well, but adds additional overhead to reinitialize the printer and transformer between each file.

How we enforce this invariant is that one of first things we do in `transformNodes` is to clean up all `emitNode` entries on the parse tree. Its highly possible that we already have a bug in this implementation due to the fact we call `transformNodes` first for .js output and then for .d.ts output, erasing any `emitNode` entries we need for JS emit before we ever get to the point of printing. Its likely that we just don't have a test case that covers this scenario currently, and that we might see breaks in our RWC tests due to this.

Reply:
I don't think it's possible to have a situation where it matters, unless a transformer sets an emit node on a node in a different file.

Reply:
As I said, we probably already have situations where this matters since the `transformNodes` call for the declaration emit basically already disposes the transform result for the javascript emit. We just probably don't have a test for that case. The most likely causes for a parse tree node having an `emitNode` revolve around custom comment or source map positions associated with identifiers. Even if we don't have an issue today, its such an obscure interaction that when we do have an issue it will be very hard to track down.

I'd be tempted to change the line in transformNodes that ensures the parse tree is disposed to instead perform a `Debug.assert` if there are undisposed parse tree nodes so that we can catch these errors more readily. That would enforce that you `.dispose()` a transform result before you reuse the parse tree in another call to `transformNodes`.

Reply:
Yeah, that just seems like a bit of an obscure API with a bunch of implicit invariants that can't be enforced in the typesystem. It's probably better if we can just keep emit nodes in a side table associated with each transform/transform result; especially since it's super easy just to accidentally call `dispose` before printing, for example, which would have silently messed up your print. (And the table already exists globally fir the implementation of dispose in the first place, iirc) If they're stored in the transform result, then dispose is just no longer necessary. (And we'd no longer be making parse tree nodes polymorphic on the presence of emit nodes, which is probably good and has potential to outweigh the extra indirection, just like inverting the checker's cache structure experimentally did)

Reply:
We could use a side table only for the parse tree nodes, since we wouldn't want to hold onto intermediate nodes created by transforms that end up elided by a later transform as there's no point in holding onto information we don't need. The problem is that we would have to have a significant breaking change to the public API for transforms to support that (since right now those functions just look at `.emitNode`). Its much easier and less likely to break the API if we just ensure that we finish the JS transformation before we start the declaration one. For single file emit that's not a problem because we work at the Bundle level anyways, but for multi-file output we would have to transform and emit a single file at a time. That is a much more palatable change (and one that I've looked at once or twice in the past when analyzing performance), since we won't need the `emitNode` data between multiple files for multi-file output.